pinot-core/src/main/java/org/apache/pinot/core/query/selection/SelectionOperatorUtils.java

```diff
@@ -182,10 +187,12 @@ public static Pair<DataSchema, int[]> getResultTableDataSchemaAndColumnIndices(Q
 
     // No order-by expression
     // NOTE: Order-by expressions are ignored for queries with LIMIT 0.
+    boolean isSelectStar = queryContext.getQueryOptions() != null
+        && QueryOptionsUtils.isSelectStarQuery(queryContext.getQueryOptions());
     List<OrderByExpressionContext> orderByExpressions = queryContext.getOrderByExpressions();
     if (orderByExpressions == null || queryContext.getLimit() == 0) {
       // For 'SELECT *', use the server response data schema as the final results data schema.
-      if ((numSelectExpressions == 1 && selectExpressions.get(0).equals(IDENTIFIER_STAR))) {
+      if ((numSelectExpressions == 1 && selectExpressions.get(0).equals(IDENTIFIER_STAR) || isSelectStar)) {
```

Comment:
If this called on the broker, select(*) is always expanded. Can we just use the isSelectStart condition check then?